Add support for TS Teletext subtitles

cdongieux opened the pull request:
This PR adds support for Teletext subtitles in TS streams.

Some readings:
- EN 300 472 - V1.4.1 - Digital Video Broadcasting (DVB); Specification for conveying ITU-R System B Teletext in DVB bitstreams
- EN 300 468 - V1.16.1 - Digital Video Broadcasting (DVB); Specification for Service Information (SI) in DVB systems
- ETS 300 706 - Enhanced Teletext specification (good luck with this one)

You can test this PR with this sample file, its a dump of a French television channel stream "NRJ12" with 2 teletext tracks : https://drive.google.com/file/d/19ZYsoeUfH85gEilQkaAdLbPhC4CxhDEh/view?usp=sharing
You can also compare the result in ExoPlayer with VLC which handles Teletext subs too.

Let me know if you need further info :)

Commit:
Add support for TS Teletext subtitles

icbaker commented:
Thanks for sending this - it's quite a large change, so I'm afraid you'll have to bear with me while I find the time to review it.

icbaker commented:
Do you have any test clips containing teletext subtitles that you could add to this PR? We could use them in:
* [`TsExtractorTest`](https://github.com/google/ExoPlayer/blob/dev-v2/library/extractor/src/test/java/com/google/android/exoplayer2/extractor/ts/TsExtractorTest.java) to ensure the track is extracted
* Unit tests for the new `DvbTeletextDecoder`

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/extractor/ts/DvbTeletextReader.java:
```diff
@@ -0,0 +1,100 @@
+      if (teletextInfo.type == 0x02) { // Teletext subtitle page
+        formatBuilder.setRoleFlags(C.ROLE_FLAG_SUBTITLE);
+      } else if (teletextInfo.type == 0x05) { // Teletext subtitle page for hearing impaired people
+        formatBuilder.setRoleFlags(C.ROLE_FLAG_DESCRIBES_MUSIC_AND_SOUND);
```
I think in this case the flags should **also** contain ROLE_FLAG_SUBTITLE:

```java
formatBuilder.setRoleFlags(C.ROLE_FLAG_SUBTITLE | C.ROLE_FLAG_DESCRIBES_MUSIC_AND_SOUND);
```

cdongieux replied:
Don't you think this would be better?
```
formatBuilder.setRoleFlags(C.ROLE_FLAG_CAPTION | C.ROLE_FLAG_DESCRIBES_MUSIC_AND_SOUND);
```
Because this type of subs adds information about music and sound, on top of dialogs.

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/extractor/ts/TsExtractor.java:
```diff
@@ -749,6 +753,24 @@ private EsInfo readEsInfo(ParsableByteArray data, int length) {
+          streamType = TS_STREAM_TYPE_TELETEXT;
+          dvbTeletextInfos = new ArrayList<>();
+          while (data.getPosition() < positionOfNextDescriptor) {
+            String teletextLanguage = data.readString(3).trim();
```
Can you use the `readString` overload that takes a `Charset` and pass [`Charsets.ISO_8859_1`](https://guava.dev/releases/snapshot-jre/api/docs/com/google/common/base/Charsets.html#ISO_8859_1)? The version you're using atm assumes UTF-8

From ETSI EN 300 468 - V1.16.1 - 6.2.43 (emphasis mine):
> **ISO_639_language_code:** This 24-bit field contains the 3 character ISO 639-2 [36] language code of the language of the teletext. Both ISO 639-2/B and ISO 639-2/T may be used. **Each character is coded into 8 bits according to ISO/IEC 8859-1 [17] and inserted in order into the 24-bit field.**

cdongieux replied:
Done in 0037e8917c8fc3f5510d3e56e520f0f9cb8ef6ff

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/extractor/ts/TsExtractor.java:
```diff
@@ -749,6 +753,24 @@ private EsInfo readEsInfo(ParsableByteArray data, int length) {
+            byte[] initializationData = new byte[2];
+            data.readBytes(initializationData, 0, 2);
+            byte typeAndMag = initializationData[0];
+            int type = typeAndMag >> 3;
```
I think this is dodgy if `typeAndMag` has its MSB set to 1, i.e. it's negative in Java's signed view of the world. It looks like the Java sign-propagation and expansion to `int` causes problems here:

```java
byte typeAndMag = (byte) 0x80;  // Java thinks this is -128

// Java expands the -128 byte (0x80) to -128 int (0xFFFF_FFC0) and then shifts this by
// 3 to give type=0xFFFF_FFF8 (-8). I think the code below would expect type to be equal
// to 0x10 (16).
int type = typeAndMag >> 3;  
```

You can avoid gotchas like this by using [`ParsableBitArray`](https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/util/ParsableBitArray.html) and only reading the number of bits you care about. I'd suggest doing that, and reading the bytes directly from `ParsableByteArray` into it using the `readBytes(ParsableBitArray, int)` overload.

cdongieux replied:
Done in cb97b5637c0906fb72a10a6ffc2e90f1db98da94

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/extractor/ts/TsPayloadReader.java:
```diff
@@ -111,6 +117,27 @@ public DvbSubtitleInfo(String language, int type, byte[] initializationData) {
+  final class DvbTeletextInfo {
+
+    public final String language;
+    public final int type;
```
This can be a byte, right? (it's max 5 bits afaict)

cdongieux replied:
Done in cb97b5637c0906fb72a10a6ffc2e90f1db98da94

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/extractor/ts/TsExtractor.java:
```diff
@@ -749,6 +753,24 @@ private EsInfo readEsInfo(ParsableByteArray data, int length) {
+              continue;
+            }
+            dvbTeletextInfos.add(
+                new DvbTeletextInfo(teletextLanguage, type, initializationData));
```
optional: i'd consider flipping this, to put the `.add()` call inside the `if()` block.

Also, nit: Can you move the comments out of the conditional statement and onto their own line inside the if-block?

```java
if (type == 0x02 || type == 0x05) {
  // Only 'Teletext subtitle page' (0x02) and 'Teletext subtitle page for hearing 
  // impaired people' (0x05) are supported.
  dvbTeletextInfos.add(
      new DvbTeletextInfo(teletextLanguage, type, initializationData));
}
```

(this may be irrelevant if you remove the conditional completely)

cdongieux replied:
Done in cb97b5637c0906fb72a10a6ffc2e90f1db98da94

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/extractor/ts/DvbTeletextReader.java:
```diff
@@ -0,0 +1,100 @@
+import java.util.List;
+
+/** Parses DVB Teletext subtitle data and extracts individual frames. */
+public final class DvbTeletextReader implements ElementaryStreamReader {
```
The name of this class makes it sound like it handles all aspects DVB teletext data, but the javadoc only mentions subtitles.

Should it be renamed to make it clear it only supports subtitles? e.g. `DvbTeletextSubtitleReader`

cdongieux replied:
Done in 4578299b5cfe65018085a24537c6e29e676a1cb2

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/extractor/ts/TsExtractor.java:
```diff
@@ -749,6 +753,24 @@ private EsInfo readEsInfo(ParsableByteArray data, int length) {
+            int type = typeAndMag >> 3;
+            if (type != 0x02 /* Teletext subtitle page */
+                && type != 0x05 /* Teletext subtitle page for hearing impaired people */) {
+              // other types are not subtitles, ignore them
```
What do you think about handling all values of `teletext_type` here, and moving the 'is it 0x02 or 0x05?' logic into `DvbTeletext(Subtitle)Reader` (which actually already has conditional statements for these two types)

I don't think there's a problem with handling all teletext packets the same here - but I could be wrong.

It feels better to keep all the 'type checking/filtering' in one place if possible.

cdongieux replied:
Done in cb97b5637c0906fb72a10a6ffc2e90f1db98da94

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/extractor/ts/TsPayloadReader.java:
```diff
@@ -111,6 +117,27 @@ public DvbSubtitleInfo(String language, int type, byte[] initializationData) {
+    /**
+     * @param language The ISO 639-2 three-letter language code.
+     * @param type The Teletext type.
+     * @param initializationData The magazine number and page number.
```
I'm not sure the current implementation follows this javadoc? On L763 you read all 16 bits into the two bytes of this array, which covers 5 bits of type, 3 bits of mag number and 8 bits of page number. But AFAICT you never modify the first byte of this array to 'remove' the type bits.

This might also be clearer if you switch to `ParsableBitArray` as suggested above.

If you do this, I would consider having explicit fields in this type for magazine number and page number, and building the `initializationData` array later.

(oh, it looks like in `DvbTeletextDecoder` you've compensated for this by only reading the lowest 3 bits of the first byte: `int magazineNumber = data.readUnsignedByte() & 0x07;` - which does suggest this javadoc is being ignored?)

cdongieux replied:
Fixed in cb97b5637c0906fb72a10a6ffc2e90f1db98da94

icbaker replied:
What do you think about this suggestion?

> If you do this, I would consider having explicit fields in this type for magazine number and page number, and building the initializationData array later.

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextDecoder.java:
```diff
@@ -0,0 +1,36 @@
+   */
+  public DvbTeletextDecoder(List<byte[]> initializationData) {
+    super("DvbTeletextDecoder");
+    ParsableByteArray data = new ParsableByteArray(initializationData.get(0));
```
nit: i think you can avoid this allocation by using guava's [`UnsignedInts.toInt`](https://guava.dev/releases/snapshot-jre/api/docs/com/google/common/primitives/UnsignedBytes.html#toInt(byte)) directly on the `initializationData` array.

cdongieux replied:
Done in cb97b5637c0906fb72a10a6ffc2e90f1db98da94

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextParser.java:
```diff
@@ -0,0 +1,513 @@
+import java.util.List;
+
+/** Parses {@link Cue}s from a DVB Teletext subtitle bitstream. */
+/* package */ final class DvbTeletextParser {
```
Does this need to be its own class? It kinda feels like it could be in-lined into `DvbTeletextDecoder`

cdongieux replied:
Well, I applied exactly the same logic as the existing `DvbParser`/`DvbDecoder` classes.

icbaker replied:
Ack - let's leave as-is.

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextParser.java:
```diff
@@ -0,0 +1,513 @@
+import java.util.Collections;
+import java.util.List;
+
+/** Parses {@link Cue}s from a DVB Teletext subtitle bitstream. */
```
Please add a reference to the relevant spec here

I haven't fully reviewed the rest of this file yet since I'm not sure which spec I should be consulting

cdongieux replied:
Done in 69f49d5388fbdc6972d7ca7517171fad3617d4de

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextParser.java:
```diff
@@ -0,0 +1,513 @@
+  private static final int TTX_ROWS = 24;
+  private static final int CHAR_HEIGHT = 21;
+
+  private static final int[] TXT_COLORS = {
```
Don't use abbreviations like TXT unless it's a literal reference to something in the spec

cdongieux replied:
Done in 69f49d5388fbdc6972d7ca7517171fad3617d4de

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextParser.java:
```diff
@@ -0,0 +1,513 @@
+    int offset = 0;
+    int len = size;
+
+    int dataIdentifier = data[offset++] & 0xFF;
```
This might be better with a `ParsableByteArray`? 

If you want to avoid the allocation you could use a scratch instance as a field, and reset it with `data` each time.

cdongieux replied:
Done in 69f49d5388fbdc6972d7ca7517171fad3617d4de

wazerstar commented:
> Do you have any test clips containing teletext subtitles that you could add to this PR? We could use them in:
> 
>     * [`TsExtractorTest`](https://github.com/google/ExoPlayer/blob/dev-v2/library/extractor/src/test/java/com/google/android/exoplayer2/extractor/ts/TsExtractorTest.java) to ensure the track is extracted
> 
>     * Unit tests for the new `DvbTeletextDecoder`


3 samples, you click over to the right side of the page "Download"

https://cyberfile.is/Fbo

icbaker commented:
> 3 samples, you click over to the right side of the page "Download"
> 
> https://cyberfile.is/Fbo

This seems to just open repeated NSFW pop-ups, please upload these somewhere less dubious (I haven't been able to download them yet).

It also looks like these are going to be too large to be added to git and used for automated testing, but may be useful for manual testing. In order to for us to accept this PR we also need some very small sample files (<1MB each) (that can have any copyright assigned under the CLA you've signed) into a git commit and put directly in this PR (in this directory: https://github.com/google/ExoPlayer/tree/dev-v2/testdata/src/test/assets/media/ts).

It might also be good to write a `DvbTeletextDecoderTest` - I recently started a `Cea608DecoderTest` that might be a useful starting point: https://github.com/google/ExoPlayer/blob/dev-v2/library/extractor/src/test/java/com/google/android/exoplayer2/text/cea/Cea608DecoderTest.java

wazerstar commented:
> > 3 samples, you click over to the right side of the page "Download"
> > https://cyberfile.is/Fbo
> 
> This seems to just open repeated NSFW pop-ups, please upload these somewhere less dubious (I haven't been able to download them yet).
> 
> It also looks like these are going to be too large to be added to git and used for automated testing, but may be useful for manual testing. In order to for us to accept this PR we also need some very small sample files (<1MB each) (that can have any copyright assigned under the CLA you've signed) into a git commit and put directly in this PR (in this directory: https://github.com/google/ExoPlayer/tree/dev-v2/testdata/src/test/assets/media/ts).
> 
> It might also be good to write a `DvbTeletextDecoderTest` - I recently started a `Cea608DecoderTest` that might be a useful starting point: https://github.com/google/ExoPlayer/blob/dev-v2/library/extractor/src/test/java/com/google/android/exoplayer2/text/cea/Cea608DecoderTest.java


Just as a note I'm trying to help with samples here, I'm not the owner of this topic.

With all respect for me I cannot see it's possible to have a raw sample under ~15-20mb with .ts format that includes teletext subtitles, I tried, unless somebody else knows a way and can contribute with ?

Here you go.

gdrive link.
https://drive.google.com/file/d/1MOeoHCdQ4M9HIuItL-_vkfaYwwcFMsrt/view?usp=sharing

Commit:
Use ParsableByteArray.readString() with Charsets.ISO_8859_1 to extract Teletext language

Commit:
Move supported Teletext types filtering in DvbTeletextReader, enhance descriptor parsing in TsExtractor

Commit:
Rename DvbTeletext* classes to DvbTeletextSubtitle*

Commit:
Use ParsableByteArray to parse Teletext subtitle bitstream

Commit:
Add sample file containing 2 Teletext subtitles tracks for TsExtractorTest

cdongieux commented:
> Do you have any test clips containing teletext subtitles that you could add to this PR? We could use them in:
> 
> * [`TsExtractorTest`](https://github.com/google/ExoPlayer/blob/dev-v2/library/extractor/src/test/java/com/google/android/exoplayer2/extractor/ts/TsExtractorTest.java) to ensure the track is extracted
> * Unit tests for the new `DvbTeletextDecoder`

I have added a sample TS file containing Teletext subtitle only (no audio, no video), and I updated TsExtractorTest with a new test.

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/extractor/ts/DvbTeletextSubtitleReader.java:
```diff
@@ -0,0 +1,112 @@
+  @Override
+  public void createTracks(ExtractorOutput extractorOutput, TrackIdGenerator idGenerator) {
+    for (int i = 0; i < outputs.length; i++) {
+      DvbTeletextInfo teletextInfo = teletextInfos.get(i);
```
This looks dodgy, because above in the constructor you only add items to `outputs` for a subset of `teletextInfos`, so these indices might not match up.

What about using an `ArrayList` for `outputs` instead? So above in the constructor you just store `teletextInfos` into a field and instantiate an empty `outputs` list, then move all the filtering logic into this method and add items to `outputs` as you go.

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/extractor/ts/DvbTeletextSubtitleReader.java:
```diff
@@ -0,0 +1,112 @@
+      if (type == SUBTITLE_PAGE) {
+        formatBuilder.setRoleFlags(C.ROLE_FLAG_SUBTITLE);
+      } else {
+        formatBuilder.setRoleFlags(C.ROLE_FLAG_DESCRIBES_MUSIC_AND_SOUND);
```
As discussed in a [previous comment](https://github.com/google/ExoPlayer/pull/10588#discussion_r993382189), should this be OR'd with `C.ROLE_FLAG_CAPTION`?

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/extractor/ts/TsExtractor.java:
```diff
@@ -749,6 +754,23 @@ private EsInfo readEsInfo(ParsableByteArray data, int length) {
+          while (data.getPosition() < positionOfNextDescriptor) {
+            ParsableBitArray scratch = new ParsableBitArray(new byte[5]);
+            data.readBytes(scratch, 5);
+            String teletextLanguage = scratch.readBytesAsString(3, Charsets.ISO_8859_1).trim();
```
nit: you can probably read the string from the `ParsableByteArray` and only use the `ParsableBitArray` for the next fiddly bit?

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/extractor/ts/TsExtractor.java:
```diff
@@ -749,6 +754,23 @@ private EsInfo readEsInfo(ParsableByteArray data, int length) {
+            dvbTeletextInfos.add(
+                new DvbTeletextInfo(
+                    teletextLanguage, (byte) type,
+                    new byte[] { (byte) magazineNumber, (byte) pageNumber })
```
Consider moving the byte casting up to where the locals are declared and assigned (i.e. declare them as type `byte`). And please use Guava's [`UnsignedBytes.checkedCast`](https://guava.dev/releases/snapshot-jre/api/docs/com/google/common/primitives/UnsignedBytes.html#checkedCast(long)).

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+      0x8F, 0x4E, 0x4F, 0x8F, 0x4E, 0x0E, 0x8F, 0x4E,
+  };
+
+  private static final int[] SPECIAL_CHARS_IDX;
```
I don't see the word 'special' used in ETS 300 706 to describe characters or character sets.

What about sticking with the terminology in the spec and naming this something about "national option sub-set"?

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+  private static void handleSpecialChars(byte[] data, int offset, int len, int lang) {
+    boolean mosaic = false;
+    for (int i = offset; i < offset + len; ++i) {
+      int c = data[i] & 0xFF;
```
imo this is clearer if spelled [`UnsignedBytes.toInt(data[i])`](https://guava.dev/releases/snapshot-jre/api/docs/com/google/common/primitives/UnsignedBytes.html#toInt(byte))

Same elsewhere

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+  private static final char[][] SPECIAL_CHARS = {
+      /* for latin-1 font */
+      /* English (100%) */
+      {0, '£', '$', '@', '«', '½', '»', '¬', '#', '­', '¼', '¦', '¾', '÷'},
```
it looks like this is derived from Table 36 in section 15.6.2 in ETS 300 706, but I'm not sure I understand why the leading `0` is needed (which afaict isn't present in the table).

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+  };
+
+  private static final int[] SPECIAL_CHARS_IDX;
+  private static final char[][] SPECIAL_CHARS = {
```
This charset stuff is really hard to follow...

What about using a Guava ImmutableTable<Integer, Byte, Character> to directly replicate table 36 in the spec. That might remove lots of these arrays and make the usage a bit easier to reason about?

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+  };
+
+  // 7bit with 1 bit Odd parity
+  private static final byte[] PARITY_TABLE = {
```
I know this pattern already exists in `Cea608Decoder` - but I'm not sure this is better than just using [`Integer.bitCount`](https://developer.android.com/reference/java/lang/Integer#bitCount(int))? I was actually considering switching `Cea608Decoder` to use that - since imo it's more obviously correct to a future reader (analysing this array by hand for correctness is very tricky).

Something like:

```java
private static void checkParity(byte[] data, int offset, int len) {
  for (int i = 0; i < len; ++i) {
    int a = data[offset + i] & 0xFF;
    if (Integer.bitCount(UnsignedBytes.toInt(data[offset + i])) % 2 == 0) {
      Log.w(TAG, "parity error at " + (offset + i));
    }
    data[offset + i] &= 0x7F;
  }
}
```

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+    for (int i = 0; i < len; ++i) {
+      int a = data[offset + i] & 0xFF;
+      if ((PARITY_TABLE[a] & 32) == 0) {
+        Log.w(TAG, "parity error at " + (offset + i));
```
shouldn't decoding fail here rather than just logging and continuing?

Same below for unrecoverable hamming codes

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+
+  private static final int[] CHARSET_INDEXES = {0, 4, 2, 6, 1, 5, 3, 7};
+
+  private static void checkParity(byte[] data, int offset, int len) {
```
This method also removes the parity bit, I think that should be made clear in the name - e.g. `checkAndRemoveParityBit`

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+          break;
+      }
+    }
+    return new ArrayList<>(cues);
```
Can you switch this to `ImmutableList.copyOf`?

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+
+  @Nullable
+  TeletextLine processTeletextSubtitleField(ParsableByteArray data, int len) {
+    if (len != 0x2C) {
```
Section 7.1 suggests a teletext packet is 45 bytes, but this is checking `len == 44` - why the discrepancy?

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+    }
+
+    data.setPosition(data.getPosition() - 2);
+    revertBytes(data.getData(), data.getPosition(), len);
```
Is this needed because of this bit of section 7.1?
> In each byte, the bits are numbered 1 to 8 (LSB to MSB), and are normally transmitted LSB first.

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+  }
+
+  // revert bits order from LSB first to MSB first
+  private static void revertBytes(byte[] data, int offset, int len) {
```
I don't think 'revert' is the right word here

Maybe reverse? But that might suggest reversing `data` rather than reversing each byte

["Bit endian"](https://en.wikipedia.org/wiki/Endianness#Bit_endianness[citation_needed]) is probably a good phrase that's less ambiguous, so maybe `correctBitEndianness` and reference section 7.1 of the ETS 300 706 spec from the javadoc of this method.

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+  public List<Cue> decode(byte[] data, int size) {
+    scratch.reset(data, size);
+
+    int dataIdentifier = scratch.readUnsignedByte();
```
I couldn't work out where in ETS 300 706 this is defined - could you reference the relevant section from a code comment here?

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+    return new ArrayList<>(cues);
+  }
+
+  private Cue buildCue(TeletextLine line) {
```
nit: please move this below `processTeletextSubtitleField` since that is called first in `decode` above.

icbaker replied:
It's not obvious to me that we want one `Cue` per line

What about encoding multiple lines into a `Cue` using `\n` to separate the lines?

This is what we do in `Cea608Decoder`:
https://github.com/google/ExoPlayer/blob/r2.18.1/library/extractor/src/main/java/com/google/android/exoplayer2/text/cea/Cea608Decoder.java#L974

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+  }
+
+  @Nullable
+  TeletextLine processTeletextSubtitleField(ParsableByteArray data, int len) {
```
can this be private?

icbaker replied:
The ETS 300 706 spec seems (I think?) to refer to this as a 'teletext packet' rather than a 'field' - would it be clearer to use that wording here too?

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+      return null;
+    }
+
+    data.setPosition(data.getPosition() + 1);
```
this is probably more clearly spelled as `data.skipBytes(1)`

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+    return updatePage(page, packetNumber);
+  }
+
+  private TeletextLine updatePage(@NonNull TeletextPage page, int row) {
```
This is a slightly confusing name for a method that mostly (?) creates a `TeletextLine` instance.

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+
+    len -= 2;
+
+    if (magazine != magazineNumber) {
```
please don't use textually distinct but semantically identical identifiers like this, it's very hard for a reader to follow what the difference between `magazine` (which is a number) and `magazineNumber` is.

In this case it's probably clearer to use `this` to disambiguate.

You could also rename the intance field to `selectedMagazine` or similar? Since that seems to be how it operates (we only keep packets with the same magazine number)

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+      return null;
+    }
+
+    int rowOffset = 0;
```
looks like this is never updated from 0?

Maybe just in-line it below on L467 with a `/* parameterName= */` comment? Or remove the parameter completely since `updateLine` is only called from one place anyway

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+      len -= 8;
+    }
+
+    if (page.pageNumber != pageNumber) {
```
imo this would also be clearer if the field was called `selectedPageNumber` - does that work?

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+    if (packetNumber == 0) {
+      // ETSI EN 300 706 - 7.1.4 - Packet types: Page header
+      int pageNumber = hamming84Decode(data.readUnsignedByte(), data.readUnsignedByte());
+      if (pageNumber == 0xFF) {
```
I'm not sure I understand why this implies we need to return `null` here?

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+      if (pageNumber == 0xFF) {
+        return null;
+      }
+      page.pageNumber = pageNumber;
```
Not sure this if this is possible, just a thought:

It seems we don't really ever need/expect `this.page.pageNumber != this.pageNumber` - so what about checking `this.pageNumber == pageNumber` here, and returning early if they're different - maybe you could then completely remove the `Page.pageNumber` field? 

What about exiting early here so

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+      page.pageNumber = pageNumber;
+
+      // ETSI EN 300 706 - 9.3.1 - Page header
+      data.skipBytes(1);
```
nit: can you add a comment here that we're skipping 'Subcode S1' - just to help a reader reacquaint themselves.

```java
data.skipBytes(1);  // Subcode S1
```

Same below

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+      // ETSI EN 300 706 - 9.3.1 - Page header
+      data.skipBytes(1);
+      int s2AndC4 = hamming84Decode(data.readUnsignedByte());
+      page.erase = (s2AndC4 & 0x08) != 0;
```
Table 2 describes C4 as being in location 'Byte 9, bit 8'. `& 0x08` would seem to select either bit 3 or 4 (if counting from LSB, 0 or 1 indexed), or bit 4 or 5 (if counting from MSB, 0 or 1 indexed)

Have I misunderstood what's happening here?

EDIT: Ah, maybe the spec is referring to bit positions before the Hamming protection is removed?

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+      page.inhibitDisplay = (c7C10 & 0x8) != 0;
+      int c11C14 = hamming84Decode(data.readUnsignedByte());
+      page.charset = CHARSET_INDEXES[c11C14 >> 1];
+      len -= 8;
```
it looks like you're duplicating the position/limit tracking that `ParsableByteArray` already provides. What about just using the `position` and `bytesLeft` from there?

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+      return null;
+    }
+
+    if (packetNumber >= 1 && packetNumber <= 24 && page.inhibitDisplay) {
```
nit: move the `page.inhibitDisplay` check to the front, i think it reads more naturally

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+    }
+
+    if (packetNumber >= 1 && packetNumber <= 24 && page.inhibitDisplay) {
+      page.erase = true;
```
This doesn't quite match the definition of Inhibit Display in Table 2:
> Data addressed to rows 1 to 24 is not to be displayed.

It doesn't say anything about erasing existing data - I would assume we could just return here before decoding these packets instead?

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+  }
+
+  private TeletextLine updatePage(@NonNull TeletextPage page, int row) {
+    TeletextLine dr = new TeletextLine(row, page.erase, page.lines[row]);
```
I think it's a little confusing to have mutable array references shared between `TeletextLine` and `TeletextPage` like this. Is it necessary?

I wonder if we need the `TeletextLine` type at all, and we could just encode everything in `TeletextPage`?

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+    boolean box = false;
+    boolean newBox = false;
+
+    for (int i = 0; i < line.line.length; ++i) {
```
can we rename something to get rid of this `line.line`? It's a bit confusing to think of a line 'having' a line inside it.

(this may be moot if you get rid of `TeletextLine` as floated below)

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+
+    for (int i = 0; i < line.line.length; ++i) {
+      int c = line.line[i] & 0xFF;
+      if (c < 0x20) {
```
At this point I was lost in the spec again - can you add another signpost comment?

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+        } else if (c == 0xb) {
+          newBox = true;
+        } else if (c == 0xd) {
+          size = (float) CHAR_HEIGHT / TELETEXT_HEIGHT * 2;
```
I think size is either equal to `CHAR_HEIGHT / TELETEXT_HEIGHT` or half that (or is it double? not sure about operator precedence between `/` and `*`), and `CHAR_HEIGHT` and `TELETEXT_HEIGHT` are otherwise unused.

What about removing the `CHAR_HEIGHT` and `TELETEXT_HEIGHT` constants and just having a `float TEXT_SIZE = 21f / 576f` constant that you can divide/multiply by 2 if needed?

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+        charsetName = StandardCharsets.ISO_8859_1.name();
+      } else {
+        charsetName = "ISO-8859-1";
+      }
```
You can use Guava [`Charsets`](https://guava.dev/releases/snapshot-jre/api/docs/com/google/common/base/Charsets.html) on any build

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+    } catch (UnsupportedEncodingException e) {
+      latin1Line = "";
+    }
+    SpannableStringBuilder stringBuilder = new SpannableStringBuilder(latin1Line);
```
nit: no need to use `SpannableStringBuilder` (which lets you modify both the string content and the spans) - [`SpannableString`](https://developer.android.com/reference/android/text/SpannableString) should be fine.

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+      box = newBox;
+    }
+
+    String latin1Line;
```
nit: no need to name this `latin1Line`, can't it just be `line` or `cueText`?

You could even in-line it into the `SpannableString(Builder)` constructor on L369 imo to avoid having to name it at all.

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+    SpannableStringBuilder stringBuilder = new SpannableStringBuilder(latin1Line);
+    stringBuilder.setSpan(new ForegroundColorSpan(color), 0, latin1Line.length(),
+        Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
+    stringBuilder.setSpan(new BackgroundColorSpan(0x00000000), 0, latin1Line.length(),
```
It looks like this adds a transparent background - does that have any effect?

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java:
```diff
@@ -0,0 +1,511 @@
+      0xffff00ff, // Alpha Magenta
+      0xff00ffff, // Alpha Cyan
+      0xffffffff, // Alpha White
+  };
```
For lots of these you can use [Color](https://developer.android.com/reference/android/graphics/Color) constants which will be self-documenting and you can remove the comments.

zolsen23 commented:
Hi @cdongieux, I'm really looking forward to this improvement :) I did some testing on my side with a MPEG-TS file and I can see that the subtitles appear for a few seconds and then disappear completely. I have attached a video of what I see. [https://www.dropbox.com/s/6rx12iifbms6kq7/device-2022-11-29-092900.mp4?dl=0](url)

This is the video file details:
`General
ID                                       : 1 (0x1)
Complete name                            : C:\Users\zolsen\Videos\239.1.1.16.ts
Format                                   : MPEG-TS
File size                                : 2.47 GiB
Duration                                 : 1 h 59 min
Overall bit rate mode                    : Variable
Overall bit rate                         : 2 951 kb/s

Video
ID                                       : 161 (0xA1)
Menu ID                                  : 785 (0x311)
Format                                   : MPEG Video
Format version                           : Version 2
Format profile                           : Main@Main
Format settings                          : CustomMatrix / BVOP
Format settings, BVOP                    : Yes
Format settings, Matrix                  : Custom
Format settings, GOP                     : Variable
Codec ID                                 : 2
Duration                                 : 1 h 59 min
Bit rate mode                            : Variable
Bit rate                                 : 2 288 kb/s
Maximum bit rate                         : 2 332 kb/s
Width                                    : 720 pixels
Height                                   : 576 pixels
Display aspect ratio                     : 16:9
Active Format Description                : Full frame 16:9 image
Frame rate                               : 25.000 FPS
Standard                                 : PAL
Color space                              : YUV
Chroma subsampling                       : 4:2:0
Bit depth                                : 8 bits
Scan type                                : Progressive
Scan order                               : Top Field First
Compression mode                         : Lossy
Bits/(Pixel*Frame)                       : 0.221
Time code of first frame                 : 12:55:39:22
Time code source                         : Group of pictures header
GOP, Open/Closed                         : Open
Stream size                              : 1.92 GiB (78%)

Audio #1
ID                                       : 81 (0x51)
Menu ID                                  : 785 (0x311)
Format                                   : MPEG Audio
Format version                           : Version 1
Format profile                           : Layer 2
Codec ID                                 : 3
Duration                                 : 1 h 59 min
Bit rate mode                            : Constant
Bit rate                                 : 256 kb/s
Channel(s)                               : 2 channels
Sampling rate                            : 48.0 kHz
Frame rate                               : 41.667 FPS (1152 SPF)
Compression mode                         : Lossy
Delay relative to video                  : -488 ms
Stream size                              : 220 MiB (9%)
Language                                 : English

Audio #2
ID                                       : 82 (0x52)
Menu ID                                  : 785 (0x311)
Format                                   : MPEG Audio
Format version                           : Version 1
Format profile                           : Layer 2
Codec ID                                 : 3
Duration                                 : 1 h 59 min
Bit rate mode                            : Constant
Bit rate                                 : 256 kb/s
Channel(s)                               : 2 channels
Sampling rate                            : 48.0 kHz
Frame rate                               : 41.667 FPS (1152 SPF)
Compression mode                         : Lossy
Delay relative to video                  : -488 ms
Stream size                              : 220 MiB (9%)
Language                                 : English
Language, more info                      : Visual impaired commentary
Editorial classification                 : Visual impaired commentary
Mix type                                 : Independent

Text
ID                                       : 41 (0x29)-801
Menu ID                                  : 785 (0x311)
Format                                   : Teletext Subtitle
Language                                 : English`

wazerstar commented:
@cdongieux You got an update on this code?

cdongieux commented:
@wazerstar sorry no, I didn't have time to check on this.

wazerstar commented:
@google-admin @icbaker @cdongieux can we get this ticket migrated to https://github.com/androidx/media

icbaker commented:
Closing all PRs on this deprecated project. We are now unable to merge PRs from here. If you would like us to consider this change again please file a new PR on the media3 project: https://github.com/androidx/media/pulls
